src/makeStyles.tsx

```diff
@@ -48,44 +58,81 @@ export function createMakeStyles<Theme>(params: { useTheme: () => Theme }) {
             return function useStyles(params: Params) {
                 const theme = useTheme();
 
-                const { cx, css } = useCssAndCx();
+                const { css, cx } = useCssAndCx();
+
+                const cache = useTssEmotionCache();
 
                 return useMemo(() => {
+                    const refClassesCache: Record<string, string> = {};
+
+                    const refClasses = new Proxy<
```

Comment:
>  explicit generic arguments to the makeStyles invocation which is quite annoying  

Passing `{ "ref": [ "a", "b", "c" ] }` if anything is more verbose that "a" | "b" | "c". The only thing that pain me is to have to explicitly provide void as the `Param` argument.  `makeStyle<void, "a" | "b" | "c">` 😔

>  it also provides no safety at all for non-typescript use  

Making explicitly provide the `"ref": [ "a", "b", "c" ]` provide no safety to the JS users. If anything it's one more place where they can screw up.  In JSS they where used to `$ruleName` without having to provide an extra parameter. They expect the same with TSS.  

There is also two extra argument in favor of using type only param instead of runtime param:  

- For `withStyles()`  (API that I personally don't use but that is important to many non the less) combined with MUI components, it's a game changer:  
![demo_withStyles](https://user-images.githubusercontent.com/6702424/136705025-dadfff08-7d9a-49f7-8696-533ca38ec38f.gif)
- It make writing [the codemod](https://github.com/garronej/tss-react/issues/32) for automatic migration from v4's `makeStyles` and `withStyles` achievable.

Reply:
Well as mentioned before already, `withStyles` also has a third argument to pass options so that should not be an issue

Reply:
Sorry I provided the wrong gif.  
I already explained it to you. Why would you require user to explicitly pass in an information we don't need and we already have?  
It would be a fundamentally flawed design.  
In `withStyles`, it's not the user that get to decide the rules name. They are defined by the type of the`classes` props.